libc/startup/linux/CMakeLists.txt

```diff
@@ -27,18 +70,24 @@ function(add_startup_object name)
   )
 endfunction()
 
+check_cxx_compiler_flag("-r" LIBC_LINKER_SUPPORTS_RELOCATABLE)
+
+if(NOT LIBC_LINKER_SUPPORTS_RELOCATABLE)
+  message(STATUS "Skipping startup for target architecture ${LIBC_TARGET_ARCHITECTURE}: linker does not support -r")
+  return()
+endif()
+
 if(NOT (EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/${LIBC_TARGET_ARCHITECTURE}))
   message(STATUS "Skipping startup for target architecture ${LIBC_TARGET_ARCHITECTURE}")
   return()
 endif()
 
 add_subdirectory(${LIBC_TARGET_ARCHITECTURE})
 
-add_startup_object(
+# TODO: factor out crt1 into multiple objects
```

Comment:
I don't follow what this TODO is implying. Can you expand on what your intent is? Sorry, it's not obvious to me.

Reply:
Sure.

The `do_start` function is actually architecture-independent. With this patch, we can have something like:

```
+ x86_64
  - start.cpp (provides _start, set_tls_ptr, init_tls)
+ aarch64
  - start.cpp (provides _start, set_tls_ptr, init_tls)
+ riscv
  - start.cpp (provides _start, set_tls_ptr, init_tls)
+ do_start.cpp (or some other proper name, provides `do_start` function)
```

Then `do_start.cpp.o` and `{arch}/start.cpp.o` can be merged and generate the final `crt1.o`.

Reply:
I see what you mean wrt. do_start. Consider expanding further on your comment, unless that's something you intend to pursue immediately after this patch?

Reply:
Yes, I hope to separate `do_start` out of `start.cpp` immediately after this patch. 

Some other related changes:
- `x86_64` does not have `do_start` for now, it can be changed to
```c++
extern "C" void _start() {
  app.args = reinterpret_cast<LIBC_NAMESPACE::Args *>(
        reinterpret_cast<uintptr_t *>(__builtin_frame_address(0)) + 1);
  __asm__ __volatile__("andq $0xfffffffffffffff0, %rsp\n\t");
  __asm__ __volatile__("andq $0xfffffffffffffff0, %rbp\n\t");
  do_start()
}
```
- for `riscv`, move the following to `_start`:
   ```c++
     LIBC_INLINE_ASM(".option push\n\t"
                  ".option norelax\n\t"
                  "lla gp, __global_pointer$\n\t"
                  ".option pop\n\t");
    ```

Reply:
https://github.com/llvm/llvm-project/pull/75717 is another related change. The purpose of #75717 is to define `LIBC_NAMESPACE::app` symbol regardless it is in the full build or the overlay mode; but the patch also reduced "duplicated" definition of the same symbol.